### PR TITLE
Rowgeno

### PR DESCRIPTION
**Pull request by cseed**
@tomwhite @jkeebler Added a parquet_genotypes flag to VDS.write:

```
import hail
hc = hail.HailContext()
vds = hc.import_vcf('sample.vcf')
vds.count()
vds.write('sample.pq.vds', overwrite=True, parquet_genotypes=True)
```

Then in the Spark shell:

```
scala> val df = spark.read.parquet("sample.pq.vds/rdd.parquet")
scala> df.printSchema()
root
 |-- variant: struct (nullable = true)
[variant and annotation fields elided]
 |-- gs: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- gt: integer (nullable = true)
 |    |    |-- ad: array (nullable = true)
 |    |    |    |-- element: integer (containsNull = true)
 |    |    |-- dp: integer (nullable = true)
 |    |    |-- gq: integer (nullable = true)
 |    |    |-- px: array (nullable = true)
 |    |    |    |-- element: integer (containsNull = true)
 |    |    |-- fakeRef: boolean (nullable = true)
 |    |    |-- isDosage: boolean (nullable = true)
```

I added correctness tests, but no performance testing on the Hail side yet.

Note, the `px` field is the `PL` in the case of sequence data and 16-bit fixed-point dosages in the case of array data (See `Gentoype` for more details.)  We know if we have dosage or not globally (`VariantMetadata.isDosage`), so I can customize the resulting schema in v2.

Finally, I'm seeing `containsNull = true` here, but I set it to `containsNull = false` when I constructed the schema programatically.  Spark/Parquet seem to be consistently ignoring my non-missing hints.  Have you seen this before?  Any idea why it is happening?

From `Genotype.schema`:

```
  def schema: DataType = StructType(Array(
    StructField("gt", IntegerType),
    StructField("ad", ArrayType(IntegerType, containsNull = false)),
    StructField("dp", IntegerType),
    StructField("gq", IntegerType),
    StructField("px", ArrayType(IntegerType, containsNull = false)),
    StructField("fakeRef", BooleanType, nullable = false),
    StructField("isDosage", BooleanType, nullable = false)))
```

### CODE REVIEW

**Comment by tpoterba** [comment]
why is `containsNull` true on the spark-shell readout?

**Comment by tpoterba** [comment]
Yeah, I saw `containsNull=false`, that's why I asked.  Very strange.

**Comment by cseed** [comment]
Old but relevant: http://stackoverflow.com/questions/30004295/dataframe-save-sqlcontext-load-loses-nullable-status-of-schema

**Comment by cseed** [comment]
Possible Parquet can't store this information?

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 1]
```diff
@@ -266,6 +281,43 @@ abstract class Genotype extends Serializable {
...
+    if (r.isNullAt(3))
+      None
+    else
+      Some(r.getInt(3))
+
+  def px: Option[Array[Int]] =
+    if (r.isNullAt(4))
+      None
+    else
+      Some(r.getSeq(4).toArray)
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/Genotype.scala]: this could be really inefficient for genotype functions that use PLs a lot

**Comment by tomwhite** [comment]
It looks like the Parquet schema in the file stores the `nullable = false` (as a `required` field), but it's ignored when being read back. I ran the following on a Spark 2 shell (updated from the Stack Overflow question, note that the Parquet handling has been rewritten since then, see https://issues.apache.org/jira/browse/SPARK-9095):

```scala
scala> import org.apache.spark.sql._
import org.apache.spark.sql._

scala> import org.apache.spark.sql.types._
import org.apache.spark.sql.types._

scala> val sqlContext = new org.apache.spark.sql.SQLContext(sc)
warning: there was one deprecation warning; re-run with -deprecation for details
sqlContext: org.apache.spark.sql.SQLContext = org.apache.spark.sql.SQLContext@625f5712

scala> val schema = StructType(Seq(StructField("foo", IntegerType, false)))
schema: org.apache.spark.sql.types.StructType = StructType(StructField(foo,IntegerType,false))

scala> val df1 = sqlContext.createDataFrame(sc.parallelize(Array(Row(1))), schema)
df1: org.apache.spark.sql.DataFrame = [foo: int]

scala> df1.printSchema
root
 |-- foo: integer (nullable = false)


scala> df1.write.parquet("temp.df1")
                                                                                
scala> val df2 = sqlContext.read.parquet("temp.df1")
df2: org.apache.spark.sql.DataFrame = [foo: int]

scala> df2.printSchema
root
 |-- foo: integer (nullable = true)
```

Then

```bash
parquet-tools schema part-r-00000-94aa6aa3-4799-4b78-9717-5397c8e983f9.snappy.parquet
message spark_schema {
  required int32 foo;
}
```

Which shows that the field is `required`, not `optional`.

Also

```bash
parquet-tools meta part-r-00000-94aa6aa3-4799-4b78-9717-5397c8e983f9.snappy.parquet
creator:     parquet-mr version 1.5.0-cdh5.7.0 (build ${buildNumber}) 
extra:       org.apache.spark.sql.parquet.row.metadata = {"type":"struct","fields":[{"name":"foo","type":"integer","nullable":false,"metadata":{}}]} 

file schema: spark_schema 
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
foo:         REQUIRED INT32 R:0 D:0
```

Which shows that the Spark SQL schema is stored in the Parquet metadata.

So, looks like a bug. Is this causing correctness/perf problems?

**Comment by cseed** [comment]
> So, looks like a bug. Is this causing correctness/perf problems?

We aren't seeing correctness problems.  Performance, I dunno.  You'd hope Spark would take advantage of non-missingness, but this makes me think it isn't.  I doubt the only barrier is tracking the information at the type level.

**Pull request merged**